### PR TITLE
libindex: new fetcher implementation

### PR DESCRIPTION
**Pull request by hdonnay**
This moves the fetcher implementation into the libindex package, so that
a Libindex object de-duplicates and reference counts all the layers it
has fetched.

### CODE REVIEW

**Comment by hdonnay** [comment]
Possibly also addresses the same issue as #416

**Comment by crozzy** [comment]
I think the concept is exactly right, we want an object whose lifetime matches with the external dependency lifetime (i.e. the PVC or storage device attached to the pod, or the "lease" over the host operating system), that allows for more control over the storage and not have to trust all the various moving parts to clean up after each other.

The random suffix approach was a band-aid over the problem. The one concern would be the locking slowing things down but the small operations within the lock mean the latency is probably negligible.

Go does provide this [sync.Map](https://pkg.go.dev/sync#Map) but I'm not sure if this would qualify as `multiple goroutines read, write, and overwrite entries for disjoint sets of keys`.

**Comment by hdonnay** [comment]
I suppose it probably is disjoint for most cases, but I think the sync.Map's lack of compare-and-swap would make it hard to use correctly. I think if lock contention becomes and issue, we can swap out the `Mutex`+`map[string]int` for an `RWMutex`+`map[string]*int` and use atomic operations on the returned pointers, with the mutex being used to lock the map *only*, instead of the map and its contents.

But if you're happy with this strategy as a first whack, I'll polish this up and undraft it.

**Comment by crozzy** [comment]
Yeah, I'm happy :+1:

File: libindex/fetcher.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+// Close marks all the layers' backing files as unused.
+//
+// This method may actually delete the backing files.
+func (f *FetchProxy) Close() error {
+	var err error
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	for _, digest := range f.clean {
+		_, e := f.f.decRef(digest)
```
Comment (crozzy) [inline libindex/fetcher.go:342]: Can we consider `fp` (and possibly `fa`) so lines like this (`fp.fa.decRef`) are a little easier to read?

**Review by crozzy — COMMENTED** [review]
As mentioned before, I think this is the right approach to fetching, just a couple of comments but overall LGTM

File: libindex/fetcher.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+func (f *FetchArena) decRef(digest string) (int, error) {
+	f.mu.Lock()
+	f.rc[digest]--
+	ct := f.rc[digest]
+	if ct == 0 {
+		delete(f.rc, digest)
+	}
+	f.mu.Unlock()
+	if ct == 0 {
```
[the remark this answers is not in the document]
Comment (hdonnay) [reply libindex/fetcher.go:69]: I was trying to avoid a call to `Remove` while holding the lock, but I think that might be getting too fancy with it.

**Pull request merged**